src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ImplementIDisposableCorrectly.cs

```diff
@@ -106,13 +114,506 @@ public abstract class ImplementIDisposableCorrectlyAnalyzer : DiagnosticAnalyzer
                                                                              DiagnosticSeverity.Warning,
                                                                              isEnabledByDefault: false,
                                                                              description: s_localizableDescription,
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
+                                                                             helpLinkUri: HelpLinkUri,
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(IDisposableReimplementationRule, FinalizeOverrideRule, DisposeOverrideRule, DisposeSignatureRule, RenameDisposeRule, DisposeBoolSignatureRule, DisposeImplementationRule, FinalizeImplementationRule, ProvideDisposeBoolRule);
 
         public override void Initialize(AnalysisContext analysisContext)
         {
+            analysisContext.RegisterCompilationStartAction(
+                context =>
+                {
+                    INamedTypeSymbol disposableType = WellKnownTypes.IDisposable(context.Compilation);
+                    if (disposableType == null)
+                    {
+                        return;
+                    }
+
+                    var disposeInterfaceMethod = disposableType.GetMembers(DisposeMethodName).Single() as IMethodSymbol;
+                    if (disposeInterfaceMethod == null)
+                    {
+                        return;
+                    }
+
+                    INamedTypeSymbol garbageCollectorType = context.Compilation.GetTypeByMetadataName(GarbageCollectorTypeName);
+                    if (garbageCollectorType == null)
+                    {
+                        return;
+                    }
+
+                    var suppressFinalizeMethod = garbageCollectorType.GetMembers(SuppressFinalizeMethodName).Single() as IMethodSymbol;
+                    if (suppressFinalizeMethod == null)
+                    {
+                        return;
+                    }
+
+                    var analyzer = new Analyzer(context.Compilation, disposableType, disposeInterfaceMethod, suppressFinalizeMethod);
```

Comment:
In this case initialization can be done in constructor, because it is private class. However, in general I prefer to separate object construction and initialization because of several reasons:

- What if initialization is asynchronous
- What if object must be disposed even after failed initialization

So I did the same separation here.

Also in my opinion the code is more readable like this. Now object is created using `ctor` and then action is called `Initialize`. In case there is `ctor` only, the object is created without any action. It might not be clear, why it is created in the first place.